File: 116. Populating Next Right Pointers in Each Node.py

# Definition for a Node.
class Node:
    def __init__(self, val: int = 0, left: 'Node' = None, right: 'Node' = None, next: 'Node' = None):
        self.val = val
        self.left = left
        self.right = right
        self.next = next

class Solution:
    def connect(self, root: 'Node') -> 'Node':
        if not root:
            return
        def BFS(go):
            next_q = []
            for n in range(len(go)):
                # print(go[n].val,end=' ')
                if go[n].left or go[n].right:
                    next_q.append(go[n].left)
                    next_q.append(go[n].right)
                if n != len(go) - 1:
                    go[n].next = go[n + 1]
            if next_q:
                BFS(next_q)

        BFS([root])
        return root

File: 116. Populating Next Right Pointers in Each Node.c
/**
 * Definition for a Node.
 * struct Node {
 *     int val;
 *     struct Node *left;
 *     struct Node *right;
 *     struct Node *next;
 * };
 */

void setNext(struct Node* root){
    if (!root->left){
        return;
    }
    root->left->next = root->right;
    if (!root->next){
        //The most right root case.
        root->right->next = NULL;
    }else{
        //The other root case.
        root->right->next = root->next->left;
    }
    setNext(root->left);
    setNext(root->right);
}


struct Node* connect(struct Node* root) {
    if (!root){
        return NULL;
    }
    setNext(root);
    return root;
}
